pkg/scheduler/framework/preemption/preemption.go

```diff
@@ -401,6 +460,91 @@ func (ev *Evaluator) prepareCandidate(ctx context.Context, c Candidate, pod *v1.
 	return nil
 }
 
+// prepareCandidateAsync triggers a goroutine for some preparation work:
+// - Evict the victim pods
+// - Reject the victim pods if they are in waitingPod map
+// - Clear the low-priority pods' nominatedNodeName status if needed
+// The Pod won't be retried until the goroutine triggered here completes.
+//
+// See http://kep.k8s.io/4832 for how the async preemption works.
+func (ev *Evaluator) prepareCandidateAsync(c Candidate, pod *v1.Pod, pluginName string) {
+	metrics.PreemptionVictims.Observe(float64(len(c.Victims().Pods)))
+
+	// Intentionally create a new context, not using a ctx from the scheduling cycle, to create ctx,
+	// because this process could continue even after this scheduling cycle finishes.
+	ctx, cancel := context.WithCancel(context.Background())
+	errCh := parallelize.NewErrorChannel()
+	preemptPod := func(index int) {
+		victim := c.Victims().Pods[index]
+		if err := ev.PreemptPod(ctx, c, pod, victim, pluginName); err != nil {
+			errCh.SendErrorWithCancel(err, cancel)
+		}
+	}
+
+	ev.mu.Lock()
+	ev.preempting.Insert(pod.UID)
+	ev.mu.Unlock()
+
+	logger := klog.FromContext(ctx)
+	go func() {
+		startTime := time.Now()
+		result := metrics.GoroutineResultSuccess
+		defer metrics.PreemptionGoroutinesDuration.WithLabelValues(result).Observe(metrics.SinceInSeconds(startTime))
+		defer metrics.PreemptionGoroutinesExecutionTotal.WithLabelValues(result).Inc()
+		defer func() {
+			if result == metrics.GoroutineResultError {
+				// When API call isn't successful, the Pod may get stuck in the unschedulable pod pool in the worst case.
+				// So, we should move the Pod to the activeQ.
```

Comment:
Note that it will be moved to the backoff queue, so it's one more argument why we should do this only in error case (I guess in the non-error case we'd rather prefer to move it to active directly).

Reply:
One more thought. Note that the removed low priority pods won't be able to jump again in this spot until the pod times out in the backoff queue. I guess it's not bad, as this place will stay reserved when the pod is retried.

I however have a different question. If some pod changes the nominated node name, is there an event that can wake up pods waiting for such a change? In our case, the removed low priority pods may remain unschedulable in case the preempted pod got eventually scheduled in different location, in which they couldn't due to other constraints. I know it's a bit theoretical and we have flush mechanism, but still.

Reply:
> I guess in the non-error case we'd rather prefer to move it to active directly

This is a good topic to discuss. (no need to block this PR though)
In the theory that we've had so far, backoffQ is the _**penalty**_ that the pod wasted the scheduling cycle.
So, the point would be whether we should treat the scheduling cycle that has triggered the preemption as "it was wasted" or "it was necessary".

---

> I however have a different question.

Yeah, regarding the nominated node name change (= added, cleared, changed), we don't properly handle it right now. Actually, one issue is raised, see https://github.com/kubernetes/kubernetes/issues/126700#issuecomment-2347515403.

Reply:
> Yeah, regarding the nominated node name change (= added, cleared, changed), we don't properly handle it right now. Actually, one issue is raised, see [#126700 (comment)](https://github.com/kubernetes/kubernetes/issues/126700#issuecomment-2347515403). I wanted, but couldn't make time working on this nominatedNodeName problem in general.

I was thinking about this issue as well, but I missed your last update there when you confirmed it. Sounds like something relatively easy to fix.

Reply:
If you're interested in, feel free to give it a try :)

Reply:
> If you're interested in, feel free to give it a try :)

Yeah, I can give it a try, but note that I will be OOO next week.